### PR TITLE
Feature/input dimension

### PR DESCRIPTION
**Pull request by tomalrussell**
Introduces `Spec` and `Coordinates` for defining dimensions of inputs, outputs and parameters.

Here we:
- lay the groundwork for using pandas+xarray to handle data blobs and de/serialisation - a `Spec` is closely modelled on the metadata that specifies xarray's `DataArray`.
- drop the module-global region/interval registers and removes conversion from dependencies, in favour of future `Adapter` or `Converter` nodes in the model/dataflow graph.
- refactor `project.yml` to nest scenario/narrative variants within their parent groupings (dropping the 'sets'), and use the `Spec` definition everywhere for inputs/parameters/outputs
- drop the sos/sector/scenario model builders in favour of `from_dict` classmethods to match `as_dict` for config de/serialisation
- move modules under subpackages for consistency

Follow-on work:
- update and test `http_api` and `app` to reflect changes in config and metadata
- fill out file and database implementations of `data_interface` (attend to xfailed tests)
- review test coverage for corner cases
- review documentation
- apply config and wrapper changes to nismod models

### CODE REVIEW

File: src/smif/sample_project/config/project.yml [hunk 1]
```diff
@@ -1,69 +1,85 @@
...
-    description: ''
-  name: rainfall
-narrative_sets:
-- description: Describes the evolution of technology
-  name: technology
-interval_definitions:
-- description: ''
-  filename: annual_intervals.csv
-  name: annual
+    description: Growth in UK population
```
Comment (willu47) [inline src/smif/sample_project/config/project.yml]: This looks good, but a few comments:
- Is year (timestep) assumed as a default dim?
- Mohammad uses a different term to "variants" - "replicates". Should we use that? 
- I understand the `provides` heading to contain a Spec? 

Please add a description of this new format to the docs.

Under `variants | data` why is there a list for data?

File: src/smif/model/dependency.py [hunk 2]
```diff
@@ -10,71 +13,37 @@ class Dependency(object):
...
+    source : smif.metadata.Spec
+        The source output description
+    sink_model : smif.model.Model
+        The sink model object
+    sink : smif.metadata.Spec
+        The sink input description
     """
-    def __init__(self, source_model, source, sink, function=None):
-        self.logger = getLogger(__name__)
+    def __init__(self, source_model, source, sink_model, sink):
```
[the remark this answers is not in the document]
Comment (tomalrussell) [reply src/smif/model/dependency.py]: We do refuse to create dependencies if metadata don't match, up to variable names not matching (i.e. it's okay for a model to provide 'regional_gva' to another model that requires 'rGVA' if the rest of the spec matches). 

Storing references to both source and sink spec should be little overhead for a little convenience - we could hold onto one spec and both variable names separately if that seems cleaner.

File: src/smif/sample_project/config/project.yml (hunk 1 above)
Comment (tomalrussell) [reply src/smif/sample_project/config/project.yml]: Documentation: TODO..

Timestep is currently assumed to exist in the data (or is picked up if present), but is not included in the spec because we assume that models only request data in the context of a timestep (so they don't expect it as a dimension). It does seem a little tricky - almost an optional dimension. We do see some cases where multi-timestep data is requested for the whole simulation period, e.g. (I think) for energy demand floor area data.

Variants is my best attempt at avoiding `_sets`. A `scenario` here has `variants` exactly as a `scenario_set` had `scenarios` previously. So we can have a population scenario with high/baseline/low variants, or an economic scenario with HHC/LLC/CCH/CCL... variants.  Happy to roll back or change the wording if it's not helpful.

Replicates would be good candidates for a dimension: a river flows scenario could provide 100 replicates of flows at each river gauge in Ml/day. Dims would be `['gauge', 'day', 'replicate']`. Otherwise we'd have a list of variants like `['near-future_0', 'near-future_1', ... 'far-future_0', 'far-future_1' ... ]` each with dims `['gauge', 'day']`.

Each scenario_set (scenario) `provides` a list of outputs, each defined with a `Spec`, exactly.

Each scenario (variant) must then provide data for each of those outputs. The `dict` under `data` must have a key for each variable name, and the value must be the filename where that variable data can be found.

**Comment by tomalrussell** [comment]
Merging to [`develop`](https://github.com/nismod/smif/tree/develop) for further work.

File: src/smif/__init__.py [hunk 3]
```diff
@@ -15,3 +16,6 @@
     __version__ = pkg_resources.get_distribution(__name__).version
 except:
     __version__ = 'unknown'
+
+warnings.filterwarnings("ignore", message="numpy.dtype size changed")
```
[the remark this answers is not in the document]
Comment (tomalrussell) [reply src/smif/__init__.py]: Done with https://github.com/nismod/smif/pull/203/commits/0c83621f80ad8a1c014cc2ee79ea024f6d073749

File: src/smif/metadata/coordinates.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
+"""Dimensions are always finite, and data which has a given dimension can be indexed by a
+set of coordinates
+"""
+
+
+class Coordinates(object):
```
[the remark this answers is not in the document]
Comment (tomalrussell) [reply src/smif/metadata/coordinates.py]: Done with https://github.com/nismod/smif/pull/203/commits/a80a7f3c9927f154c56d005b6c11f970b8565a31

File: src/smif/metadata/coordinates.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
+"""Dimensions are always finite, and data which has a given dimension can be indexed by a
```
[the remark this answers is not in the document]
Comment (tomalrussell) [reply src/smif/metadata/coordinates.py]: Done with https://github.com/nismod/smif/pull/203/commits/a80a7f3c9927f154c56d005b6c11f970b8565a31

File: src/smif/metadata/spec.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
+"""Data is typically multi-dimensional, a Spec provides metadata to label it.
+"""
+from collections import defaultdict
+
+from smif.metadata.coordinates import Coordinates
+
+
+class Spec(object):
```
[the remark this answers is not in the document]
Comment (tomalrussell) [reply src/smif/metadata/spec.py]: Done with https://github.com/nismod/smif/pull/203/commits/a80a7f3c9927f154c56d005b6c11f970b8565a31

File: src/smif/metadata/spec.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+                 abs_range=None, exp_range=None, unit=None, description=None):
+        self._name = name
+        self._description = description
+
+        # Coords may come as a dict, in which case dims must be provided to define order
+        if isinstance(coords, dict):
+            if dims is None:
+                msg = "Spec.dims must be specified if coords are provided as a dict, in {}"
+                raise ValueError(msg.format(self._name))
+            if sorted(dims) != sorted(coords.keys()):
```
[the remark this answers is not in the document]
Comment (tomalrussell) [reply src/smif/metadata/spec.py]: Done with https://github.com/nismod/smif/pull/203/commits/d2de46f29d00c57cbd15f9aae90b304347568e4d

File: src/smif/metadata/spec.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+        self._dims = dims
+        self._coords = coords
+
+        if dtype is None:
+            raise ValueError("Spec.dtype must be provided, in {}".format(self._name))
+        self._dtype = dtype
+
+        self._default = default
+        self._abs_range = abs_range
```
[the remark this answers is not in the document]
Comment (tomalrussell) [reply src/smif/metadata/spec.py]: Done with https://github.com/nismod/smif/pull/203/commits/b5385014c65102e59ebaa9aa5b1f16ee6d2b30fb
